src/pyodide/internal/builtin_wrappers.js

```diff
@@ -0,0 +1,88 @@
+import { default as Eval } from "pyodide-internal:eval";
+
+let evalableFuncs = undefined;
+export function setEvalableFunctions(funcs) {
+  if (evalableFuncs) {
+    return;
+  }
+  console.log("setEvalableFunctions", { funcs });
+  evalableFuncs = funcs;
+}
+
+function prepareStackTrace(error, stack) {
+  if (stack.length < 3) {
+    return false;
+  }
+  delete Error.prepareStackTrace;
+  try {
+    const funcName = stack[2].getFunctionName();
+    const fileName = stack[2].getFileName();
+    return (
+      funcName === "convertJsFunctionToWasm" &&
+      fileName === "pyodide-internal:pyodide-bundle/pyodide.asm"
+    );
+  } catch (e) {
+    console.warn(e);
+    return false;
+  }
+}
+
+function checkCallee() {
+  const origPrepareStackTrace = Error.prepareStackTrace;
+  let isOkay;
+  try {
+    Error.prepareStackTrace = prepareStackTrace;
+    isOkay = new Error().stack;
+  } finally {
+    Error.prepareStackTrace = origPrepareStackTrace;
+  }
+  if (!isOkay) {
+    console.warn("Invalid call to `WebAssembly.Module`");
+    throw new Error();
+  }
+}
+
+// Wrapper for WebAssembly that makes WebAssembly.Module work by temporarily
+// turning on eval
+const origWebAssembly = globalThis.WebAssembly;
+export const WebAssembly = new Proxy(origWebAssembly, {
+  get(target, val) {
+    const result = Reflect.get(...arguments);
+    if (val !== "Module") {
+      return result;
+    }
+    return new Proxy(result, {
+      construct() {
+        checkCallee();
+        try {
+          Eval.enableEval();
+          return Reflect.construct(...arguments);
+        } finally {
```

Comment:
This is rather unfortunate from a perf point of view. `checkCallee()` is creating an Error just to get the stack, `Reflect.construct...` has some performance overhead, etc. I have to wonder if it wouldn't be more efficient just to implement this entire piece in c++... not sure tho. We'll have to play with it a bit.